### DISCUSSION TITLE
Azure OpenAI support

### DISCUSSION THREAD

**Discussion by strayer** [olimorris/codecompanion.nvim#383] [Ideas]
I'm using Azure OpenAI and would love to open a PR that adds official support for it. As far as I know the APIs are pretty much the same besides the URLs and authentication headers. For example, this is how I'm using my gpt-4o deployment right now:

```lua
  {
    "olimorris/codecompanion.nvim",
    dependencies = {
      -- ...
    },
    config = {
      strategies = {
        chat = {
          adapter = "openai",
        },
        inline = {
          adapter = "openai",
        },
      },
      adapters = {
        openai = function()
          return require("codecompanion.adapters").extend("openai", {
            url = "https://AZURE_OPENAI_RESOURCE.openai.azure.com/openai/deployments/MODEL_DEPLOYMENT_NAME/chat/completions?api-version=2024-06-01",
            env = {
              api_key = "API_KEY",
            },
            headers = {
              ["Content-Type"] = "application/json",
              ["api-key"] = "${api_key}",
            },
          })
        end,
      },
    },
  },
```

This seems to work perfectly well. I'm just wondering if you are even interested in adding this. An easy fix would be to just add this example to the README.md for adapter configuration, but a more comfortable fix could be to extend the openai adapter to work with both Azure and the official OpenAI APIs based on different configuration parameters.

What do you think? I'd love to implement this in a PR.

**Comment by olimorris** [comment]
Would absolutely welcome this. I've created an [ADAPTERS.md](https://github.com/olimorris/codecompanion.nvim/blob/main/doc/ADAPTERS.md) file which walks through the adapter creation process. You can also see in the xAI and Copilot adapters how I piggy-backed onto the OpenAI one. Expect it will be relatively straightforward and we can avoid writing specific tests too.

**Reply by strayer** [reply]
Awesome! I'll take a look and hopefully open a first draft PR (or even a final one) until the weekend!
